Secure boot state checking for Classic and Core no matter if the enable or disable (New)

seankingyang opened the pull request:
## Description

Adds symmetric secure boot state validation driven by the `has_secure_boot` manifest entry:

- `miscellanea/secure-boot-enabled` — runs when `manifest.has_secure_boot == 'True'`, fails unless secure boot is actually enabled.
- `miscellanea/secure-boot-disabled` — runs when `manifest.has_secure_boot == 'False'`, fails unless secure boot is actually disabled.

Exactly one of the pair runs per machine, so a system whose real secure boot state contradicts the manifest can no longer pass certification silently.

Detection methods in `check_secure_boot_state.py`:

- **UEFI systems (x86 and ARM with real EBBR)**: reads the `SecureBoot` EFI variable directly — the same source `mokutil` parses and the same method the existing `boot_mode_test.py` uses — so no external tool is required (mokutil is not seeded on Core/IoT images).
- **FIT-image systems (ARM boards booting u-boot FIT images)**: inspects the booted FIT kernel image with `dumpimage`; a signature indicates a verified-boot setup. This checks signature presence, not bootloader enforcement — documented in the job purpose.
- **Method selection**: `auto` (default) uses the `SecureBoot` variable when it exists, otherwise the FIT image. Some ARM platforms expose `/sys/firmware/efi` without full EBBR support, making their `SecureBoot` variable untrustworthy — no runtime probe can detect that, so those platforms pin the method explicitly with `SECURE_BOOT_CHECK_METHOD=fit` in the Checkbox configuration (passed to the jobs via `environ:`) or `--method fit`. ARM EFI platforms without EBBR *and* without a FIT image cannot be checked generically; those need a vendor-specific method.

### Rework after review (2026-07)

Following @pieqq's review the implementation was rewritten in place; the diff is now **+482 lines** (was +2444):

| | before | after |
|---|---|---|
| `check_secure_boot_state.py` | 957 lines: custom logger class, hand-rolled subprocess timeout, ABC + enum registry, `hostnamectl`-based Core detection | 182 lines of plain functions reusing `checkbox_support.snap_utils.system` and stdlib `logging` / `subprocess` |
| tests | 1438 lines / 94 tests | 256 lines / 22 behavior-focused tests |
| CLI | `--enable` / `--disabled` | positional `enabled\|disabled` |
| test plans | added as blockers to `misc-client-cert-automated` (reaching all client certs) | scoped to the three `client-cert-iot-*` plans only |

## Resolved issues

https://warthogs.atlassian.net/browse/OEMQA-6211

## Documentation

The job `_purpose` fields document the detection methods and the FIT-signature limitation.

## Tests

- `python3 -m unittest tests.test_check_secure_boot_state` — 22 tests OK
- `flake8` and `black -l 79` clean
- Real UEFI host with secure boot disabled: `check_secure_boot_state.py disabled` → `PASS`, exit 0; `check_secure_boot_state.py enabled` → clear `FAIL`, exit 1; state agrees with `mokutil --sb-state`
- Earlier DUT runs of the job pair on ARM Core (FIT/u-boot) and UEFI machines are in the comments below; re-verification of the rewritten script on the ARM boards is planned

seankingyang committed:
Add secure boot state checking improvements

seankingyang committed:
fix soem issue and change all error to customerror

seankingyang committed:
Add the unit test

seankingyang committed:
Improve the unit test coverage

seankingyang committed:
fit the py3.5 sytnax and flake8 format

rickwu666666 commented:
I've tried with my project which is `ARM64`/`secure boot enable`/`FIT image`/`ubuntu core20`. And everything works as expected.
pass result:
```
==============[ Running job 2 / 3. Estimated time left: 0:00:01 ]===============
-------[ Verify secure boot is enabled when manifest indicates support ]--------
ID: com.canonical.certification::miscellanea/secure-boot-enabled
Category: com.canonical.plainbox::miscellanea
... 8< -------------------------------------------------------------------------
Detected Ubuntu Core via hostnamectl
Initialized FITImageSecureBootChecker with hostfs_prefix='/var/lib/snapd/hostfs', ubuntu_variant='UbuntuVariant.CORE'
Detected Ubuntu variant: UbuntuVariant.CORE
Running in snap environment, using hostfs prefix: /var/lib/snapd/hostfs
No boot kernel found in common locations: ['/boot/vmlinuz', '/boot/vmlinuz-5.4.0-1079-shiner-devel', '/boot/kernel.img', '/boot/Image']
Search patterns (hostfs prioritized): ['/var/lib/snapd/hostfs/snap/*/current/kernel.img', '/var/lib/snapd/hostfs/snap/*/*/kernel.img', '/var/lib/snapd/hostfs/var/lib/snapd/seed/systems/*/kernel/kernel.img', '/var/lib/snapd/hostfs/run/mnt/ubuntu-boot/uboot/ubuntu/*/kernel.img', '/var/lib/snapd/hostfs/boot/uboot/*/kernel.img', '/snap/*/current/kernel.img', '/snap/*/*/kernel.img', '/var/lib/snapd/seed/systems/*/kernel/kernel.img', '/run/mnt/ubuntu-boot/uboot/ubuntu/*/kernel.img', '/boot/uboot/*/kernel.img']
Using snap kernel: /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img
Checking image: /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img
Checking signature of image: /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img
Image /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img is signed
Image /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img is signed
=== Secure Boot State Check ===
Checker: FIT Image (dumpimage)
Current state: enabled
PASS: Secure boot is enabled
------------------------------------------------------------------------- >8 ---
Outcome: job passed
==============[ Running job 3 / 3. Estimated time left: 0:00:00 ]===============
-----[ Verify secure boot is disabled when manifest indicates no support ]------
ID: com.canonical.certification::miscellanea/secure-boot-disabled
Category: com.canonical.plainbox::miscellanea
Job cannot be started because:
 - resource expression "manifest.has_secure_boot == 'False'" evaluates to false
Outcome: job cannot be started
```
fail result
```
==============[ Running job 2 / 3. Estimated time left: 0:00:01 ]===============
-------[ Verify secure boot is enabled when manifest indicates support ]--------
ID: com.canonical.certification::miscellanea/secure-boot-enabled
Category: com.canonical.plainbox::miscellanea
Job cannot be started because:
 - resource expression "manifest.has_secure_boot == 'True'" evaluates to false
Outcome: job cannot be started
==============[ Running job 3 / 3. Estimated time left: 0:00:00 ]===============
-----[ Verify secure boot is disabled when manifest indicates no support ]------
ID: com.canonical.certification::miscellanea/secure-boot-disabled
Category: com.canonical.plainbox::miscellanea
... 8< -------------------------------------------------------------------------
Detected Ubuntu Core via hostnamectl
Initialized FITImageSecureBootChecker with hostfs_prefix='/var/lib/snapd/hostfs', ubuntu_variant='UbuntuVariant.CORE'
Detected Ubuntu variant: UbuntuVariant.CORE
Running in snap environment, using hostfs prefix: /var/lib/snapd/hostfs
No boot kernel found in common locations: ['/boot/vmlinuz', '/boot/vmlinuz-5.4.0-1079-shiner-devel', '/boot/kernel.img', '/boot/Image']
Search patterns (hostfs prioritized): ['/var/lib/snapd/hostfs/snap/*/current/kernel.img', '/var/lib/snapd/hostfs/snap/*/*/kernel.img', '/var/lib/snapd/hostfs/var/lib/snapd/seed/systems/*/kernel/kernel.img', '/var/lib/snapd/hostfs/run/mnt/ubuntu-boot/uboot/ubuntu/*/kernel.img', '/var/lib/snapd/hostfs/boot/uboot/*/kernel.img', '/snap/*/current/kernel.img', '/snap/*/*/kernel.img', '/var/lib/snapd/seed/systems/*/kernel/kernel.img', '/run/mnt/ubuntu-boot/uboot/ubuntu/*/kernel.img', '/boot/uboot/*/kernel.img']
Using snap kernel: /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img
Checking image: /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img
Checking signature of image: /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img
Image /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img is signed
Image /var/lib/snapd/hostfs/snap/hbt-imx-kernel/current/kernel.img is signed
=== Secure Boot State Check ===
Checker: FIT Image (dumpimage)
Current state: enabled
FAIL: Secure boot is not disabled (current state: SecureBootState.ENABLED)
------------------------------------------------------------------------- >8 ---
Outcome: job failed
```

rickwu666666 reviewed (dismissed):
LGTM

tomli380576 reviewed (changes requested):
lgtm overall just small things that can help simplify enum usage

seankingyang committed:
Update providers/base/bin/check_secure_boot_state.py

Co-authored-by: Zhongning Li <60045212+tomli380576@users.noreply.github.com>

seankingyang commented on providers/base/bin/check_secure_boot_state.py:
```diff
@@ -0,0 +1,957 @@
+    def is_supported(cls) -> bool:
+        """
+        Check if this secure boot method is supported on the system.
+    
+        Returns:
```
```suggestion

        Returns:
```

seankingyang committed:
Update providers/base/bin/check_secure_boot_state.py

seankingyang committed:
Correct the order for classmethod and abstractmethod

rickwu666666 reviewed (dismissed):
LGTM

seankingyang committed:
Merge branch 'main' into check_secure_boot

rickwu666666 reviewed (dismissed):
LGTM

pieqq reviewed (changes requested):
Thanks for this PR. I especially appreciate the explanations and the tests run (so this extends to @rickwu666666, thanks for testing on your project!)

I got a recent certificate request for a project where the `secure_boot_mode_*` was failing, and suggesting to use this instead, so I had a look.

One actionable request is: since the word `disabled` is used when running the command, could the opposite be `enabled` and not `enable` to be more symmetrical?

I have a few questions:

- Does this come in addition to the `secure_boot_mode_*` job, or to replace it?
- Is it intended only for IoT projects?

I had a look at the code as well. To be honest I'm worried about PR that add thousands of lines of code. Some parts are re-inventing the wheel (like all the helper functions to log things, or the implementation of a timeout), some don't follow the [guidance in place for python tests](https://github.com/canonical/checkbox/tree/main/providers#python-tests)...

I understand the need to have better secure boot checks, but I'm sure this can be done in a more simple way. Could you revisit this?

seankingyang committed:
Merge remote-tracking branch 'upstream/main' into sb-rewrite

# Conflicts:
#	providers/base/units/miscellanea/jobs.pxu
#	providers/certification-client/units/client-cert-iot-desktop-24-04.pxu
#	providers/certification-client/units/client-cert-iot-server-24-04.pxu

seankingyang committed:
Rewrite secure boot state check per review

Address pieqq's review: the previous 957-line script re-implemented
stdlib logging, subprocess timeouts, and Ubuntu Core detection that
checkbox_support already provides, and its 1438-line test suite mostly
asserted enum literals and mock wiring.

- Rewrite check_secure_boot_state.py as ~180 lines of plain functions:
  read the SecureBoot EFI variable directly on UEFI systems (the same
  source mokutil parses, so no external tool is needed, and the same
  method boot_mode_test.py uses), and check the booted FIT kernel image
  signature via dumpimage on non-UEFI systems. Reuse on_ubuntucore,
  get_kernel_snap and add_hostfs_prefix from checkbox_support.
- Replace the asymmetric --enable/--disabled flags with a positional
  enabled|disabled argument.
- Rewrite the tests as 22 behavior-focused cases (~260 lines).
- Update the job commands and purposes accordingly, and drop the
  misc-client-cert-automated blocker lines so the new jobs stay scoped
  to the IoT client cert plans; keep the manifest-only gating.

seankingyang committed:
Add secure boot detection method override

Real-device testing found ARM platforms that boot through an EFI flow
and expose /sys/firmware/efi without fully implementing EBBR: their
SecureBoot variable exists but is not trustworthy, so auto-detection
keyed on the EFI directory picks the wrong method. No runtime probe
can tell a trustworthy variable from an untrustworthy one, so the
platform class must be declarable:

- Add --method {auto,uefi,fit}, defaulting to the
  SECURE_BOOT_CHECK_METHOD environment variable (settable from the
  Checkbox configuration; both jobs pass it through via environ:)
  and then auto.
- Make auto prefer the SecureBoot variable only when it exists, then
  fall back to the FIT image check, instead of trusting the mere
  presence of /sys/firmware/efi.
- Document the remaining honest gap: ARM EFI platforms without EBBR
  and without a FIT image cannot be checked generically.
